diff --git a/mm/memory-failure.c b/mm/memory-failure.c
index cd0d559..a7ab867 100644
--- a/mm/memory-failure.c
+++ b/mm/memory-failure.c
@@ -219,10 +219,13 @@ static int kill_proc_ao(struct task_struct *t, unsigned long addr, int trapno,
  */
 void shake_page(struct page *p, int access)
 {
+        printk(KERN_ALERT "*** shake_page %lu\n",page_to_pfn(p));
 	if (!PageSlab(p)) {
+                printk(KERN_ALERT "**** calling lru_add_drain_all()\n");
 		lru_add_drain_all();
 		if (PageLRU(p))
 			return;
+                printk(KERN_ALERT "**** calling drain_all_pages()\n");
 		drain_all_pages();
 		if (PageLRU(p) || is_free_buddy_page(p))
 			return;
@@ -235,6 +238,7 @@ void shake_page(struct page *p, int access)
 	if (access) {
 		int nr;
 		do {
+                        printk(KERN_ALERT "**** shrinking slab\n");
 			nr = shrink_slab(1000, GFP_KERNEL, 1000);
 			if (page_count(p) == 0)
 				break;
diff --git a/mm/memory_hotplug.c b/mm/memory_hotplug.c
index 6f971d6..fdfb961 100644
--- a/mm/memory_hotplug.c
+++ b/mm/memory_hotplug.c
@@ -714,7 +714,7 @@ do_migrate_range(unsigned long start_pfn, unsigned long end_pfn)
 			   check this again here. */
 			if (page_count(page))
 				not_managed++;
-#ifdef CONFIG_DEBUG_VM
+#if 0 && defined(CONFIG_DEBUG_VM)
 			printk(KERN_ALERT "removing pfn %lx from LRU failed\n",
 			       pfn);
 			dump_page(page);
